External iterator API.

Metalaka opened the pull request:
Derived from: https://github.com/hoaproject/Database/pull/13#issuecomment-53678271

The main idea are:
We can define a fetching style with direction, startOffset(, style?).
And we can get an external iterator to fetch the results with the defined style.

*Not ready to merge.*

Hywan commented:
ping?
I am really interested by the PR :heart:, so please, don't give up!

Metalaka commented:
Changes/State of the PR:

*Statement Implement `IteratorAggregate`, return a `WrapperIterator`.
The iterator is designed to be rewindable, but need to be tested...

`Statement::setFetchingStyle($orientation, $offset)` definition:
default value are to `NEXT` form `FIRST`.

$orientation values can be: // same value as `\PDO::FETCH_ORI_*`
* DalStatement::FETCH_ORI_NEXT
* DalStatement::FETCH_ORI_PRIOR

$offset values can be:
* DalStatement::FETCH_OFFSET_FIRST
* DalStatement::FETCH_OFFSET_LAST
* Or an arbitrary offset (from the top of the result set, regardless of the orientation).


`Pdo\Statement` are initialized by default with cursor enabled.

I removed fetchFirst / Last / Next / Prior because it will cause unexpected behavior if used simultaneously with the iterator, thought ?
And I don't think tracking the internal result set offset is useful (with Iterator::key()) so I removed it from the PR(https://github.com/Metalaka/Database/commit/75576041d6fa92bc63a6f359a971d620f54351bc)?

Hywan commented:
Sounds good for me. Some examples?

What's next?

Metalaka commented:
```php
use Hoa\Database;

$dal = Database\Dal::getInstance(…);
$query = 'SELECT a, b FROM foo';

// One-query fetch all data (already exists)
foreach($dal->query($query)->fetchAll() as $row) {
    echo $row['a'], $row['b'], "\n";
}

// Lazy fetch all data
foreach($dal->query($query) as $row) {
    echo $row['a'], $row['b'], "\n";
}

// Fetch from last to first row of the result set.
$stmt = $dal->query($query);
$stmt->setFetchingStyle(
    Database\DalStatement::FETCH_ORI_PRIOR,
    Database\DalStatement::FETCH_OFFSET_LAST
);
foreach($stmt as $row) {
    echo $row['a'], $row['b'], "\n";
}

// Fetch from the fifth row of the result set to the first (offset 0).
$stmt = $dal->query($query);
$stmt->setFetchingStyle(
    Database\DalStatement::FETCH_ORI_PRIOR,
    4
);
foreach($stmt as $row) {
    echo $row['a'], $row['b'], "\n";
}

// Fetch the last row, ugly... , should be done in the query if possible.
$stmt = $dal->query($query);
$stmt->setFetchingStyle(
    null,
    Database\DalStatement::FETCH_OFFSET_LAST
);
$iterator = $stmt->getIterator();
$iterator->rewind();
$row = $iterator->current();
```
Maybe we should keep Statement::fetchFirst/Last for one row use case ?

Things like fetch the ten last row (first to last) isn't possible with this API.


> What's next?

Find a compatible environment to test scrollable cursors.

Hywan commented on DalStatement.php:
```diff
@@ -47,11 +47,39 @@
      * @var \Hoa\Database\IDal\WrapperStatement
      */
-    protected $statement = null;
+    protected $statement     = null;
```
We don't align attributes with constants' equal sign :wink:.

Hywan commented on DalStatement.php:
```diff
@@ -161,47 +189,31 @@ public function fetchAll()
      */
-    public function fetchLast()
+    public function setFetchingStyle($orientation = self::FETCH_ORI_NEXT,
+                                     $offset      = self::FETCH_OFFSET_FIRST)
```
CS:
```php
public function setFetchingStyle(
    $orientation = self…,
    $offset      = self…
);
```

Hywan commented on IDal/WrapperStatement.php:
```diff
@@ -82,36 +85,17 @@ public function bindParameter(
      */
-    public function fetchPrior();
+    public function setFetchingStyle($orientation = DalStatement::FETCH_ORI_NEXT,
+                                     $offset      = DalStatement::FETCH_OFFSET_FIRST);
```
CS:
```php
public function setFetchingStyle(
    $orientation = self…,
    $offset      = self…
);
```

Hywan commented on Layer/Pdo/Iterator.php:
```diff
@@ -0,0 +1,179 @@
+    public function next()
+    {
+        $this->_row = $this->getStatement()->fetch(
+            \PDO::FETCH_ASSOC,
```
Are we sure? Could be great to still have array or object results nop?

Hywan commented on Layer/Pdo/Iterator.php:
```diff
@@ -0,0 +1,179 @@
+     */
+    public function key()
+    {
+        return null;
```
Is an integer make sense? Not sure, because if we iterate backward, we don't really know what to do… So, ok for `null`, just wondering if you thought about this.

Hywan commented on Layer/Pdo/Iterator.php:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    /**
+     * Rewind the Iterator to the first element.
```
```php
- Iterator
+ iterator
```

Hywan commented on Layer/Pdo/Iterator.php:
```diff
@@ -0,0 +1,179 @@
+                \PDO::FETCH_ORI_ABS,
+                $this->_offset
+            );
+        }
```
So we can re-iterate over the same statement as expected?

Hywan replied:
Should we use `closeCursor`?

Metalaka replied:
If the driver and the database support cursors, it should.
I don't know if `closeCursor` is required, maybe depending of databases ?

Hywan replied:
If I understood correctly, `closeCursor` will allow to **reexecute** the query, and this is not what we want. However, a new iterator should be re-generate, but this is part of the user to deal with it I guess.

Hywan commented on Layer/Pdo/Statement.php:
```diff
@@ -154,63 +168,36 @@ public function fetchAll()
      */
-    public function fetchLast()
+    public function setFetchingStyle($orientation = DalStatement::FETCH_ORI_NEXT,
+                                     $offset      = DalStatement::FETCH_OFFSET_START)
```
CS.

Hywan commented:
So the final API of a statement is the following:
* `fetchAll`,
* `fetchColumn`,
* `getIterator`.

Why keeping `fetchColumn`? Not sure it is useful here.

Next? Details first :wink:. I am not very happy with the `FETCH_*` constant names. Their values are smart and correct, but I don't like the name:
* `FETCH_ORI_NEXT`, what about `FETCH_FORWARD`?
* `FETCH_ORI_PRIOR`, what about `FETCH_BACKWARD`?
* `FETCH_OFFSET_FIRST`, very correct,
* `FETCH_OFFSET_LAST`, very correct.

I understand you tried to mimic PDO naming, but… PDO naming is… It does not fit into Hoa's naming (be neutral, be neutral…).

Next next? Scrollable cursors are not supported by SQLite? According to https://bugs.php.net/39310, nop… I can ask on Twitter, thoughts?

Hywan commented:
@Metalaka Also, is it possible to support the `Countable` interface?

Hywan commented:
@Metalaka Also, is it possible (does it even make sense) to support [`Hoa\Iterator\Seekable`](http://central.hoa-project.net/Resource/Library/Iterator/Seekable.php)?

Metalaka commented:
Countable:
[pdostatement::rowcount](http://php.net/manual/en/pdostatement.rowcount.php) isn't support by all databases for SELECT query, so I don't really know what to do with that and I [removed it](https://github.com/Metalaka/Database/commit/d2dabf118802611453aaae43bdbad528e59af3bd) from the PR.

Seekable:
Somethings like below can be done, but we must deal with orientation as you pointed out [here (Iterator::key comment)](https://github.com/hoaproject/Database/pull/18#discussion_r36276883).
```php
	public function seek($position)
	{
		$this->rewind();
		while ($this->valid() && $this->key() != $position) {
			$this->next();
		}
	}
```
Moreover the API provide a similar feature:
```php
$stmt = $dal->query(…);
$stmt->setFetchingStyle(
    Database\DalStatement::FETCH_FORWARD,
    4 // The position to seek.
);
foreach($stmt as $row) { … }
or
$row = $stmt->fetchOne();
```

Statement API:
```php
// N-lines
fetchAll()
getIterator()
// One column, SELECT COUNT(a) FROM …
fetchColumn()
// One line, SELECT a, b FROM …
fetchOne($arg = FIRST | LAST) or fetchFirst() & fetchLast()
```
When returning only one column it's faster to call `fetchColumn` than instantiate a new Iterator, rewind it, check if the data is valid and then get the data.
Moreover it's clearer, same thing for a `fetchOne` method.

Hywan commented:
I would go for `fetchFirst` and `fetchLast` since `fetchOne` is not explicit. One what? One column? One line? Why is it the first or the last one? So yup, `fetchFirst` and `fetchLast` are excellent… or `fetchNth(1)` (first), `fetchNt(2)` (second) —**should we start at 0 or 1?**— and `fetchNth(-1)` for the last one, `fetchNth(-2)` for the line before the last line etc. Is it even possible?

Metalaka commented:
Index from the last result set line can't be done since we don't know the result set size in all case.
So I (re-)added fetchFirst/last.

Hywan commented:
Good :-].

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_OBJ = 5;
```
`FETCH_OBJECT`, long names are always better.

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_NUM = 3;
```
`FETCH_MAP`?

Metalaka replied:
IMO `AS_SET` seem more meaningful.

Hywan replied:
Exactly. This is a set here. :+1:

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_LAZY  = 1;
```
By reading only the name, I don't understand what it does.

Metalaka replied:
`AS_LAZY_OBJECT`, properties are (lazy) populated only at call time.

Hywan replied:
So only one object is created and data are updated for each row, is that correct? If yes, lazy object could be correct. What about `AS_REUSABLE_OBEJCT` or something similar (maybe it's a bad idea, I understand “lazy” but not sure everyone will).

Metalaka replied:
Call to `fetch(FETCH_LAZY)` will return a `\PDORow` object.
It's like with`FETCH_OBJECT` (return `\stdClass`) but there is one difference, the properties value are fetched when the user call it and not at the construct time.

Hywan replied:
Ok, so lazy is the correct adjective.

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_ASSOC = 2;
```
The name does not provide a clear difference compared to `FETCH_NUM`.

Metalaka replied:
I propose `AS_MAP` since it's key => value

Hywan replied:
Exactly, that's a map :+1:.

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_BOTH = 4;
```
Seriously, who use this :stuck_out_tongue:? For me it's more like a bad practise, thoughts? If agree, then skip it.

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_BOUND = 6;
```
The name is not meaningful too from my point of view.

Metalaka replied:
Hoa don't implement `PDO::bindColumn` so it can't be used. I remove that.

Hywan replied:
Good.

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_CLASS = 8;
```
`FETCH_AS_CLASS`?

We have to find a naming convention here. `FETCH_AS_` sounds cool, it is different from `FETCH_FORWARD` and `FETCH_BACKWARD`. We can also replace `FETCH_AS_` by simply `AS_`. In this case, we could also imagine having `FORWARD` instead of `FETCH_FORWARD`. Please, provide an example using these constants and let's see how it reads.

Metalaka replied:
I like the idea, but now it's `Database\DalStatement` which I find too much... can we put constants in an other class with shorter name ?

```php
use Hoa\Database;
$dal = Database\Dal::getInstance(…);
$stmt = $dal->query('SELECT a, b FROM foo');

$stmt->setFetchingStyle(
    Database\DalStatement::FORWARD,
    Database\DalStatement::OFFSET_FIRST,
    Database\DalStatement::AS_CLASS |
    Database\DalStatement::FROM_CLASS_NAME_IN_FIRST_COLUMN
);
foreach($stmt as $row) {
    echo $row['a'], $row['b'], "\n";
}
```

Hywan replied:
`FROM_CLASS_NAME_IN_FIRST_COLUMN`? What is it :smiley:?

Why putting constants in another class? This is part of the statement logic, so `Database\DalStatement` is good. You can write:
```php
$stmt->setFetchStyle(
    $stmt::FORWARD,
    $stmt::OFFSET_FIRST,
    $stmt::AS_CLASS
);
```
This is still valid and much more shorter. You can also use `use` to get a shorter alias.

Metalaka replied:
> FROM_CLASS_NAME_IN_FIRST_COLUMN? What is it :smiley:?

I forgot to push my work last time, will do it now :)

I don't like use custom alias, `$stmt::FORWARD` is great thanks.

Hywan commented on DalStatement.php:
```diff
@@ -47,6 +47,115 @@
+     *
+     * @var int
+     */
+    const FETCH_INTO = 9;
```
Difference with `FETCH_CLASS` (only by reading the name)?

Metalaka replied:
`AS_UPDATE_CLASS` ?

Hywan replied:
So what is the difference with `AS_LAZY_OBJECT`?

Metalaka replied:
`AS_LAZY_OBJECT` => `\PDORow`
`AS_CLASS` => user defined class or first column value as a class name.
`AS_UPDATE_CLASS` => update properties of a provided object.

Hywan replied:
`AS_UPDATE_CLASS`, what about `REUSE_OBJECT`? Or `AS_REUSABLE_OBJECT` in this case (if you would like to have `AS_` prefix, not a strong constraint).

Hywan commented on Layer/Pdo/Statement.php:
```diff
@@ -53,7 +53,28 @@ class Statement implements Database\IDal\WrapperStatement
+     *
+     * @var int
+     */
+    protected $_style      = 2;
```
What is 2? Provide an inline comment if necessary.

Hywan commented:
Now with fetch styles we will be able to use the [`Hoa\Iterator\Demultiplexer`](https://github.com/hoaproject/Iterator#demultiplexer) iterator, and that's cool :-].

Hywan commented:
Sorry if comments are only questions, but you really do a great job and I just would like to push the limits a little bit to have an almost-perfect API :-].

Hywan commented:
@Metalaka Ok sorry I commented the files instead of the diff… I am kind of lost now :-D. Can you make a list of all the constants here and points their status (accepted, need better naming etc.) with a description please?

Metalaka commented:
# Direction

```php
FORWARD                         = 0;
```
- [x] Status: accepted.

Fetch the next row in the result set.

```php
BACKWARD                        = 1;
```
- [x] Status: accepted.

Fetch the previous row in the result set.

# Offset

```php
FROM_START                      = 0;
```
- [x] Status: accepted.

Start at the first offset.

```php
FROM_END                        = -1;
```
- [x] Status: accepted.

Start at the last offset.

# Fetching style

```php
AS_LAZY_OBJECT                  = 1;
```
- [ ] Status: need validation.

Specifies that the fetch method shall return each row as an object with
variable names that correspond to the column names returned in the result
set. `AS_LAZY_OBJECT` creates the object variable names as they are
accessed.

Using `AS_LAZY_OBJECT` will return a `\PDORow` object.
It's like with`AS_OBJECT` (return `\stdClass`) but there is one difference, the
properties value are fetched when the user call it and not at construct time.

```php
AS_MAP                          = 2;
```
- [x] Status: accepted.

Specifies that the fetch method shall return each row as an array indexed
by column name as returned in the corresponding result set. If the result
set contains multiple columns with the same name, `AS_MAP` returns only a
single value per column name.

```php
AS_SET                          = 3;
```
- [x] Status: accepted.

Specifies that the fetch method shall return each row as an array indexed
by column number as returned in the corresponding result set, starting at
column 0.

```php
AS_OBJECT                       = 5;
```
- [x] Status: accepted.

Specifies that the fetch method shall return each row as an object with
property names that correspond to the column names returned in the result
set.

```php
AS_CLASS                        = 8;
```
- [x] Status: accepted.

Specifies that the fetch method shall return a new instance of the
requested class, mapping the columns to named properties in the class.
The magic `__set` method is called if the property doesn't exist in the
requested class.

```php
AS_REUSABLE_OBJECT              = 9;
```
- [x] Status: accepted.

Specifies that the fetch method shall update an existing instance of the
requested class, mapping the columns to named properties in the class.

```php
AS_DEBUG_MAP                    = 11;
```
- [ ] Status: need better naming.

Specifies that the fetch method shall return each row as an array indexed
by column name as returned in the corresponding result set. If the result
set contains multiple columns with the same name, `AS_DEBUG_MAP` returns an
array of values per column name.

# Fetching class option
These values must be used with `AS_CLASS`.

```php
CLASS_NAME_IN_FIRST_COLUMN = 262144;
```
- [ ] Status: need validation.

Determine the class name from the value of first column.

```php
SERIALIZED                 = 524288;
```
- [ ] Status: need validation.

As `AS_REUSABLE_OBJECT` but object is provided as a serialized string.
The class constructor is never called if this flag is set.

See example: [php-src/tests/fetch_serialize](https://github.com/php/php-src/blob/master/ext/pdo_mysql/tests/pdo_mysql_stmt_fetch_serialize.phpt).

```php
PROPERTIES_LATE            = 1048576;
```
- [ ] Status: need better naming.

Call the constructor before setting properties.

Metalaka commented:
Previous comment:

Hywan on https://github.com/Metalaka/Database/commit/8b6e3579fe9d1042d785a9a6cc3a6719f58e1338#commitcomment-12632321
> Where is `4`? We jump from `3` to `5`, no `6` and `7` neither.

Metalaka on https://github.com/Metalaka/Database/commit/8b6e3579fe9d1042d785a9a6cc3a6719f58e1338#commitcomment-12635966
> Since we use the value directly on [Layer/Pdo/Iterator.php#122](https://github.com/Metalaka/Database/commit/8b6e3579fe9d1042d785a9a6cc3a6719f58e1338#diff-2dac2ab217494a219dbe53c60b5b7f51R122) we must use the same value as the underlying layer (PDO)?
Or should we do a convert method to map `Hoa\Dal` values to `PDO` values ?

Hywan commented:
Hmm. Thank you for the summary. Let me think about it one or two days please. If some other @hoaproject/hoackers have some thoughts to express, feel free!

Hywan commented:
I am not sure that having `CLASS_NAME_IN_FIRST_COLUMN`, `SERIALIZED` and `PROPERTIES_LATE` are very smart. I would prefer to be relax and introduce them when needed. Thoughts?

Metalaka commented:
I agree, dealing with these options are not very clear.
Updated.

Hywan commented:
Now about the naming (because this is the last), we are very close to a consensus.

# Offset

* :white_check_mark: `FROM_START`,
* :white_check_mark: `FROM_END`.

# Direction

* :white_check_mark: `FORWARD`,
* :white_check_mark: `BACKWARD`.

# Fetching style

* :white_check_mark: `AS_MAP`,
* :white_check_mark: `AS_SET`,
* :white_check_mark: `AS_OBJECT`, we expect a `stdClass` to be generate (maybe `AS_STRUCTURE` would be better but not in the PHP culture, so let's keep `AS_OBJECT`),
* :red_circle: `AS_CLASS`, while I agree with the term, how do we specify the classname?
* :white_check_mark: `AS_REUSABLE_OBJECT`, what about choosing the best behavior for the user? *Reductio ad absurdum*: If `AS_OBJECT` generates one new instance of a `stdClass` class per fetch (per row), we understand `AS_REUSABLE_OBJECT` will reuse the same instance. It can generate some troubles in case objects are stored when looping over the rows. So we cannot choose for the user because we cannot guess the expected behavior. So that's why this is a check mark,
* :red_circle: `AS_DEBUG_MAP`, I was thinkin about `AS_MULTIMAP` (cf. multi-set), but like multi-set, multi-map is something very connotated in the literature and this is not exactly what we do here. Actually this is what we do but this is a specific result: This does not reflect our motivations when writing `AS_MULTIMAP`. `AS_DEBUG_MAP` seems good then. This is intended to be used when debugging only and it reflects our issues: We don't know exactly what data are fetched with the query in each row, using an object for the fetching is difficult, a set is not worth envisaging neither, so a map is pretty straightforward and adapted to the situation since it provides an immediate feedback, and this is exactly what we look for when debugging. However, is “as debug map” correct regarding the English language? I am not sure. This is the last point.

# Conclusion

1. We keep `AS_CLASS` if this is easy to specify the classname (maybe as the next argument of the fetching style method),
2. `AS_DEBUG_MAP` can be wrong regarding the English language, I am not sure. Thoughts?

Thanks for your work!

Hywan commented:
Note the order: Offset, direction and fetching style. This is different from your proposal: Direction, offset and fetching style. I guess this is more intuitive in this order. “Where are we?” (offset), “Where to go?” (direction), “How do we go?” (fetching style). What do you think?

Metalaka commented:
Setting the class name was [here](https://github.com/Metalaka/Database/commit/9f1138e57861ad0582a326e6b3a3c0758778cbe4#diff-2dac2ab217494a219dbe53c60b5b7f51L115), but wasn't documented, anyway I just changed the signature.
I packed fetching options into an array since there are optional parameters (class name, constructor argument, object to use/reuse).

I added two arguments to `setFetchStyle` (see [`PDOStatement::setFetchMode`](http://php.net/manual/en/pdostatement.setfetchmode.php), I don't really know if the second can be useful).
And, I reordered the arguments according to your logic.

Hywan commented:
As always, you did a great job.

Hywan commented:
Sorry to comment the diffs, once again…

Metalaka commented:
Done.

Hywan commented:
I will review the PR soon, thanks!

Hywan commented:
Damn, I thought the PR was merged… I am putting this issue on the top of the column in the board!

Hywan commented:
I still thought it was merged! Damn…… Going to merge it as soon as possible.

Hywan commented:
@Metalaka I made a PR on your fork for the rebase: https://github.com/Metalaka/Database/pull/1. This really a good job! Now we have to write tests.

@hoaproject/hoackers Any volunteer to give it a try quickly?

Metalaka commented:
https://github.com/Metalaka/Database/pull/2 merged.

Hywan commented:
So we are good? Finally?

Hywan commented:
Any idea how to provide a test suite for that?
